### PR TITLE
RTN15h*: Handle incoming DISCONNECTED while CONNECTED

### PR DESCRIPTION
**Pull request by tcard**
Implements [RTN15h*](https://docs.ably.io/client-lib-development-guide/features/#RTN15h).

### CODE REVIEW

File: ably/realtime_conn.go [hunk 1]
```diff
@@ -559,6 +566,7 @@ func (c *Connection) eventloop() {
 			if reconnecting {
 				// reset the mode
 				c.reconnecting = false
+				c.reauthorizing = true
```
Comment (gernest) [inline ably/realtime_conn.go:569]:
```suggestion
				c.reauthorizing = false
```

File: ably/realtime_conn_spec_test.go [hunk 2]
```diff
@@ -971,6 +972,268 @@ func TestRealtimeConn_RTN15c4(t *testing.T) {
...
+		ConnectionID:      "new-connection-id",
+		ConnectionDetails: &proto.ConnectionDetails{},
+	}
+
+	// Expect a UPDATED event.
+
+	var change ably.ConnectionStateChange
+	ablytest.Instantly.Recv(t, &change, stateChanges, t.Fatalf)
+
+	if change.Event != ably.ConnectionEventUpdated {
```
Comment (gernest) [inline ably/realtime_conn_spec_test.go:1229]: The test is failing with a different event here

```
--- FAIL: TestRealtimeConn_RTN15h2_Success (0.00s)
    realtime_conn_spec_test.go:1230: expected UPDATED event; got {DISCONNECTED DISCONNECTED CONNECTED [ErrorInfo :timeout code=80003 statusCode=0] See https://help.ably.io/error/80003}
```

Comment (tcard) [reply ably/realtime_conn_spec_test.go:1229]: This and similar intermittent errors should've been fixed by 324acf8.

File: ably/realtime_conn.go [hunk 3]
```diff
@@ -519,86 +527,92 @@ func (c *Connection) eventloop() {
...
 			}
-			id := c.id
+			previousID := c.id
+			c.id = msg.ConnectionID
+			c.msgSerial = 0
+			if reconnecting && mode == recoveryMode {
+				// we are setting msgSerial as per (RTN16f)
+				msgSerial, err := strconv.ParseInt(strings.Split(c.opts.Recover, ":")[2], 10, 64)
+				if err != nil {
+					//TODO: how to handle this? Panic?
```
Comment (gernest) [inline ably/realtime_conn.go:574]: We will need to figure out our logging situation at some point. I think implementing the spec is more important at the moment, so its fine thisTODO stays.

Comment (tcard) [reply ably/realtime_conn.go:574]: I just moved this code around, the TODO was already there.

Anyway, I agree we should improve logging (see #160) once the codebase becomes less of a moving target.

In this particular case, though, I think a panic is in order. A malformed msgSerial just shouldn't ever happen, and it's already panicking if it doesn't have at least two parts separated by `:`.

Comment (gernest) [reply ably/realtime_conn.go:574]: I see, if that is a case you can just as well add a panic here as part of this PR. We don't need to keep that TODO anymore.

**Pull request merged**